Added pre-processed required affinity terms to scheduler's PodInfo type.

ahg-g opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Adds processed required affinity terms to the scheduler's PodInfo. This brings 13% performance improvement for required anti-affinity benchmarks.

before
```
BenchmarkScheduling/5000Nodes/5000Pods-12                          1000  1411689   ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-12           1000  8155520   ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/5000Pods-12               1000  12699040  ns/op
BenchmarkSchedulingPreferredPodAffinity/5000Nodes/5000Pods-12      1000  11805901  ns/op
BenchmarkSchedulingPreferredPodAntiAffinity/5000Nodes/5000Pods-12  1000  12808495  ns/op
BenchmarkSchedulingNodeAffinity/5000Nodes/5000Pods-12              1000  2725989   ns/op
```
after 
```
BenchmarkScheduling/5000Nodes/5000Pods-12                          1000  1419240   ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-12           1000  7225467   ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/5000Pods-12               1000  12679068  ns/op
BenchmarkSchedulingPreferredPodAffinity/5000Nodes/5000Pods-12      1000  10958616  ns/op
BenchmarkSchedulingPreferredPodAntiAffinity/5000Nodes/5000Pods-12  1000  11353006  ns/op
BenchmarkSchedulingNodeAffinity/5000Nodes/5000Pods-12              1000  2785566   ns/op
```

**Which issue(s) this PR fixes**:
Part of #91058

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

ahg-g committed:
Added pre-processed required affinity terms to scheduler's PodInfo type.

ahg-g commented:
@Huang-Wei @alculquicondor this is ready.

alculquicondor commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -66,44 +65,27 @@ func (s *preFilterState) Clone() framework.StateData {
-		}
+	if len(updatedPod.Spec.NodeName) > 0 {
+		s.topologyToMatchedAffinityTerms.updateWithAffinityTerms(updatedPod, node, s.podInfo.RequiredAffinityTerms, multiplier)
+		s.topologyToMatchedAntiAffinityTerms.updateWithAntiAffinityTerms(updatedPod, node, s.podInfo.RequiredAntiAffinityTerms, multiplier)
```
+1 for cleaner implementation 😃

alculquicondor commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -229,35 +182,26 @@ func getTPMapMatchingExistingAntiAffinity(pod *v1.Pod, allNodes []*framework.Nod
-	affinity := pod.Spec.Affinity
-	if affinity == nil || (affinity.PodAffinity == nil && affinity.PodAntiAffinity == nil) {
-		return topologyPairsAffinityPodsMap, topologyToMatchedExistingAntiAffinityTerms, nil
+	if len(podInfo.RequiredAffinityTerms) == 0 && len(podInfo.RequiredAntiAffinityTerms) == 0 {
```
would something fail if we just returned nil? just to avoid unnecessary allocations.

ahg-g replied:
there are more than few map access in the file, needs to be verified, I don't trust the unit tests to cover all cases.

Huang-Wei replied:
It'd safe to keep the map instantiated, as some places directly access it without check if it's nil.

alculquicondor commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -337,22 +254,20 @@ func (pl *InterPodAffinity) PreFilter(ctx context.Context, cycleState *framework
 		return framework.NewStatus(framework.Error, fmt.Sprintf("failed to list NodeInfos with pods with affinity: %v", err))
 	}
 
+	podInfo := framework.NewPodInfo(pod)
```
For a follow up: Perhaps we could store something in QueuedPodInfo? This way processed terms are stored across scheduling attempts. And this is something we could also add to Spreading plugin.

alculquicondor commented on pkg/scheduler/framework/plugins/interpodaffinity/scoring.go:
```diff
@@ -48,7 +48,7 @@ func (s *preScoreState) Clone() framework.StateData {
 
 // A "processed" representation of v1.WeightedAffinityTerm.
 type weightedAffinityTerm struct {
```
I assume this is pending to be cached in a followup PR?

alculquicondor commented on pkg/scheduler/framework/v1alpha1/types.go:
```diff
@@ -65,13 +68,54 @@ func (pqi *QueuedPodInfo) DeepCopy() *QueuedPodInfo {
+		t := newAffinityTerm(pod, &term)
+		if t == nil {
+			// We get here if the label selector failed to process, this is not supposed
+			// to happen because the pod should have been validated by the api server.
```
it could still happen in malformed tests, which might lead to confusion.
Not sure if it's worth the extra machinery to propagate the error. I'll leave it to you.

alculquicondor commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -532,21 +532,19 @@ func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(pod *v1.Pod
-					podsToMove = append(podsToMove, pInfo)
-					break
-				}
+		terms := util.GetPodAffinityTerms(up.Spec.Affinity)
```
Another justification to also have the processed information in the queues.

alculquicondor commented:
Just questions and future ideas.

/lgtm
/hold
for @Huang-Wei

Huang-Wei commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -370,7 +285,7 @@ func (pl *InterPodAffinity) AddPod(ctx context.Context, cycleState *framework.Cy
 		return framework.NewStatus(framework.Error, err.Error())
 	}
-	state.updateWithPod(podToAdd, podToSchedule, nodeInfo.Node(), 1)
+	state.updateWithPod(podToAdd, nodeInfo.Node(), 1)
```
Add a TODO to remove `podToSchedule` from the interface signature?

ahg-g replied:
we can't because it is not a standard that every plugin will store the pod in their StateCycle.

Huang-Wei commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -66,44 +65,27 @@ func (s *preFilterState) Clone() framework.StateData {
-			}
-			s.topologyToMatchedAntiAffinityTerms.updateWithAntiAffinityTerms(updatedPod, node, antiAffinityTerms, multiplier)
-		}
+	if len(updatedPod.Spec.NodeName) > 0 {
```
I feel like it's not needed as `updatePod` is from snapshot which literally means assignedPod. Otherwise, L83 should be also guarded by `if len(updatedPod.Spec.NodeName) > 0`.

ahg-g replied:
sg, done.

Huang-Wei commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -432,13 +348,13 @@ func nodeMatchesAllTopologyTerms(pod *v1.Pod, topologyPairs topologyToMatchedTer
-//  topology of any "term" for the given "pod".
-func nodeMatchesAnyTopologyTerm(pod *v1.Pod, topologyPairs topologyToMatchedTermCount, nodeInfo *framework.NodeInfo, terms []v1.PodAffinityTerm) bool {
+//  topology of any of the pod's anti affinity terms.
+func nodeMatchesAnyAntiAffinityTopologies(nodeInfo *framework.NodeInfo, state *preFilterState) bool {
```
```suggestion
func nodeMatchesAnyAntiAffinityTerm(nodeInfo *framework.NodeInfo, state *preFilterState) bool {
```

As well as the comment:

```
// nodeMatchesAnyAntiAffinityTerm checks ...
```

ahg-g replied:
done.

Huang-Wei commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -415,13 +330,14 @@ func (pl *InterPodAffinity) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, state
-func nodeMatchesAllTopologyTerms(pod *v1.Pod, topologyPairs topologyToMatchedTermCount, nodeInfo *framework.NodeInfo, terms []v1.PodAffinityTerm) bool {
+//  nodeMatchesAllAffinityTopologies checks whether "nodeInfo" matches the topologies of
+// all affinity terms of the incoming pod.
+func nodeMatchesAllAffinityTopologies(nodeInfo *framework.NodeInfo, state *preFilterState) bool {
```
it checks both topologies' presence and terms' match, so prefer:

```suggestion
func nodeMatchesAllAffinityTerms(nodeInfo *framework.NodeInfo, state *preFilterState) bool {
```

ahg-g replied:
done

Huang-Wei reviewed (commented):
Some of the comments are for pre-existing code. Feel free to fix in this PR or followup PR.

alculquicondor commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -532,21 +532,19 @@ func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(pod *v1.Pod
+				podsToMove = append(podsToMove, pInfo)
+				break
 			}
 		}
```
Not really. The other affinity terms might be satisfied by other pods. But ultimately, this is just a heuristic.

Huang-Wei replied:
> The other affinity terms might be satisfied by other pods.

Good point, however it's not the current semantics yet. See the history discussion #68701. And it reminds me of a closed implementation (#68725) I tried - I gave up due to non-trivial code changes by that time. Maybe now it's a good time to re-consider "Support inter-Pod affinity to multiple Pods" as codebase is a lot neater :)

ahg-g commented:
> Some of the comments are for pre-existing code. Feel free to fix in this PR or followup PR.

Thanks Wei, fixed the function name and removed the check on NodeName being set in the update function. I amended since the changes are really minor.

Huang-Wei commented:
/lgtm

Thanks @ahg-g . Really appreciate this PR which is pretty neat!

/hold cancel

alculquicondor reviewed (commented):
still
/lgtm

Huang-Wei commented:
/retest
